framework/connection.py

```diff
@@ -0,0 +1,83 @@
+"""
+Module to connect to remote host
+"""
+
+import logging
+
+from common_ci_utils.connection import Connection
+from framework import config
+from paramiko.auth_handler import AuthenticationException, SSHException
+
+log = logging.getLogger(__name__)
+
+
+class SSHConnection:
+    """
+    A class that connects to remote host
+    """
+
+    _instance = None
+
+    def __new__(cls, *args, **kwargs):
+        if not cls._instance:
+            cls._instance = super(SSHConnection, cls).__new__(cls)
+            cls._instance._conn = None
+        return cls._instance
+
+    def __init__(self):
+        """
+        Get connection to host
+
+        Raises:
+            authException: In-case of authentication failed
+            sshException: In-case of ssh connection failed
+
+        """
+        # Initialize the connection only if it hasn't been created yet
+        if not self._conn:
+            self.host = config.ENV_DATA["noobaa_sa_host"]
+            self.user = config.ENV_DATA["user"]
+            self.password = config.ENV_DATA.get("password")
+            self.private_key = config.ENV_DATA.get("private_key")
+            try:
+                if self.private_key:
+                    self._conn = Connection(
+                        host=self.host,
+                        user=self.user,
+                        private_key=self.private_key,
+                    )
+                else:
+                    self._conn = Connection(
+                        host=self.host,
+                        user=self.user,
+                        password=self.password,
+                    )
+            except AuthenticationException as authException:
+                log.error(f"Authentication failed: {authException}")
+                raise authException
+            except SSHException as sshException:
+                log.error(f"SSH connection failed: {sshException}")
+                raise sshException
+
+    def get_connection(self):
+        """
+        Get connection to host
+
+        Returns:
+            paramiko.client: Paramiko SSH client connection to host
+
+        """
+        return self._conn
+
+    def close_connection(self):
+        """
+        Closes SSH connection
+        """
+        if self._conn:
+            self._conn.close()
+            self._conn = None
+
+
+def pytest_sessionfinish(session, exitstatus):
+    # Close the SSH connection at the end of the pytest session
+    SSHConnection().close_connection()
```

Comment:
close_connection is not a class method, so you cannot use it this way, also close_connection refer to object varaible self._conn which is not a class variable.

Reply:
done, used classmethod